### PR TITLE
NumPy 1.24

### PR DESCRIPTION
**Pull request by gmarkall**
This PR updates Numba to support NumPy 1.24, and is ready for review. Presently CI will fail due to the lack of NumPy 1.24 packages in Anaconda, but this should be resolved in time.

Each individual commit message details the changes made and their rationale - each should be reviewable as an individual change.

See testing with two other PRs:

- #8690 tests the changes in this branch with the old build matrix, demonstrating that these changes don't introduce any issues that would have been caught by the old slicing with NumPy versions.
- #8620 contains the actual history of the development of these changes, and tests every slice with a pip-installed NumPy 1.24.1.

### CODE REVIEW

**Comment by larsoner** [comment]
> Presently CI will fail due to the lack of NumPy 1.24 packages in Anaconda, but this should be resolved in time.

Just a drive-by comment/idea to potentially unblock progress here: you could use the `conda-forge` channel in the NumPy 1.24 builds temporarily to get 1.24.1 testing going

**Comment by gmarkall** [comment]
> > Presently CI will fail due to the lack of NumPy 1.24 packages in Anaconda, but this should be resolved in time.
> 
> Just a drive-by comment/idea to potentially unblock progress here: you could use the `conda-forge` channel in the NumPy 1.24 builds temporarily to get 1.24.1 testing going

Thanks for the suggestion - #8620 tests all slices against 1.24. I don't want to change the setup for this PR, because this PR is in the form I'd like reviewed and eventually merged.

**Comment by gmarkall** [comment]
(At the time I set up #8620, there were no conda-forge packages either because the first 1.24 RC had just been released, so it uses pip)

**Review by DrTodd13 — DISMISSED** [review]
Fine with the parfor related part of this PR.

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch @gmarkall, the organised commits/commit messages were very helpful in review. There's a few minor suggestions/things to look at in the review but otherwise looks good. I've checked the Azure config update and it looks like it has a good spread of NumPy versions (roughly 5 builds per supported NumPy version).

File: docs/source/user/installing.rst
Comment (gmarkall) [reply docs/source/user/installing.rst:260]: Doh!

File: numba/np/ufunc/_internal.c
Comment (gmarkall) [reply numba/np/ufunc/_internal.c:367]: Yeah, it is - can you think of a way to test this?

File: numba/np/ufunc/_internal.c
Comment (gmarkall) [reply numba/np/ufunc/_internal.c:382]: Yeah, also untested. Whilst developing I tested this by building commit 915a237 then running with NumPy 1.24, which causes this exception to raise. I tested the above exception by manually editing the code to mix up the fastcall flags. But automated testing seems a little difficult here.

File: numba/tests/test_array_methods.py
Comment (gmarkall) [reply numba/tests/test_array_methods.py:1355]: My understanding is that calling `arr.sum` with the `timedelta64` dtype kwarg always did the wrong thing in NumPy (you would not get a `timedelta64` back), so if anyone was using it, their code was not doing the "right" thing anyway.

File: numba/np/ufunc/_internal.c
Comment (stuartarchibald) [reply numba/np/ufunc/_internal.c:367]: I'm also not sure this error message is quite right, the `strncmp(crnt_name, "at", 3) != 0` predicate suggests this block executes only if the `crnt_name` is not `at`?

File: numba/tests/test_array_methods.py
Comment (stuartarchibald) [reply numba/tests/test_array_methods.py:1355]: Thanks for the explanation, this seems reasonable.

**Comment by gmarkall** [comment]
@stuartarchibald Many thanks for the review - all comments now addressed / responded to.

File: numba/np/ufunc/_internal.c
Comment (gmarkall) [reply numba/np/ufunc/_internal.c:367]: You're totally right, I've understood the logic backwards!

File: numba/np/ufunc/_internal.c
Comment (stuartarchibald) [reply numba/np/ufunc/_internal.c:382]: I agree, it's not really testable without a lot of effort to create unusual compiled objects. Is `crnt_name` null terminated? If so I think there's nothing more to do here.

**Comment by gmarkall** [comment]
@stuartarchibald Thanks for the review. I've now corrected the error message when the fastcall flags do not match, and merged `main`. If I do a quick bodge so the test always fails:

```diff
diff --git a/numba/np/ufunc/_internal.c b/numba/np/ufunc/_internal.c
index 2557c975d..94693366e 100644
--- a/numba/np/ufunc/_internal.c
+++ b/numba/np/ufunc/_internal.c
@@ -362,7 +362,7 @@ init_ufunc_dispatch(int *numpy_uses_fastcall)
             if (*numpy_uses_fastcall == -1) {
                 *numpy_uses_fastcall = crnt->ml_flags & METH_FASTCALL;
             }
-            else if (*numpy_uses_fastcall != (crnt->ml_flags & METH_FASTCALL)) {
+            else if (1 || *numpy_uses_fastcall != (crnt->ml_flags & METH_FASTCALL)) {
                 PyErr_Format(PyExc_RuntimeError,
                     "ufunc.%s() flags do not match numpy_uses_fastcall",
                     crnt_name);
```

then I get:

```
$ python -c "import numba"
RuntimeError: ufunc.accumulate() flags do not match numpy_uses_fastcall
```

so I think this change is working as intended, though it's hard to have an automated test for it as discussed previously.

File: numba/np/ufunc/_internal.c
Comment (gmarkall) [reply numba/np/ufunc/_internal.c]: Now addressed - see my main comment on the PR: https://github.com/numba/numba/pull/8691#issuecomment-1406412575

**Comment by gmarkall** [comment]
gpuci run tests

**Comment by Tatsh** [comment]
`ensure_critical_deps` needs to be updated to allow 1.24 to work.

**Comment by Tatsh** [comment]
I think you are right. I was patching 0.56.4 where the patch is needed.

https://github.com/Tatsh/tatsh-overlay/blob/master/dev-python/numba/files/numba-version-check.patch

I tested this patch with software I'm using and all tests passed once I fixed said software's numpy 1.24 issues.

File: numba/np/ufunc/_internal.c
Comment (stuartarchibald) [reply numba/np/ufunc/_internal.c]: Thanks for fixing this and testing it!

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the updates @gmarkall, I think all review comments are now addressed and this is just waiting on package availability so as to complete testing.

**Comment by gmarkall** [comment]
gpuci run tests

**Comment by gmarkall** [comment]
> What is the state of this PR?

Right now it's still the same as described above in https://github.com/numba/numba/pull/8691#pullrequestreview-1299282164:

> Thanks for the updates @gmarkall, I think all review comments are now addressed and this is just waiting on package availability so as to complete testing.

But, I'm going to try running CI again in case the packages have become available.

**Comment by gmarkall** [comment]
Ah, a merge from `main` is needed. That should kick things off anyway.

**Comment by gmarkall** [comment]
gpuci run tests

**Comment by SomeoneSerge** [comment]
Hello! Would be great to see this PR merged:)

I'm totally unfamiliar with `numba` processes, so I'll just ask: do you think there's a way you could split the release into two phases, such that you wouldn't have to wait for anaconda to get up to speed? I.e. you'd first just make `master` compatible with the new numpy, making no promises, and keep running your own CI against the last available version. And then as soon as conda has caught up you announce official compatibility?

**Comment by sklam** [comment]
@SomeoneSerge, we have decided to do something like that. We are now relying on conda-forge for testing the latest dependency. Good news is that this PR has already pass tests on linux-64 in our buildfarm. Just waiting for the rest to complete to merge this PR.

**Comment by sklam** [comment]
This PR has passed all 64-bit platforms. 32-bit platform failures are so far buildfarm issues.

**Comment by sklam** [comment]
We just need to get AzureCI to use conda-forge for np1.24

**Comment by gmarkall** [comment]
gpuci run tests

**Review by stuartarchibald — APPROVED** [review]
The changes to the public CI configuration since the last approval look as expected for use of `conda-forge` NumPy packages for NumPy version 1.24. Thanks for making this adjustment @gmarkall.

**Comment by sklam** [comment]
Public CI failed due to timeout on windows. We can safely ignore the failure given it passed on buildfarm already.

**Pull request merged**
